x/earn/keeper/vault.go

```diff
@@ -0,0 +1,196 @@
+package keeper
+
+import (
+	"github.com/cosmos/cosmos-sdk/store/prefix"
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	"github.com/kava-labs/kava/x/earn/types"
+)
+
+// GetVaultTotalSupplied returns the total balance supplied to the vault. This
+// may not necessarily be the current value of the vault, as it is the sum
+// of the supplied denom and the value may be higher due to accumulated APYs.
+func (k *Keeper) GetVaultTotalSupplied(
+	ctx sdk.Context,
+	denom string,
+) (sdk.Coin, error) {
+	vault, found := k.GetVaultRecord(ctx, denom)
+	if !found {
+		return sdk.Coin{}, types.ErrVaultRecordNotFound
+	}
+
+	return vault.TotalSupply, nil
+}
+
+// GetTotalValue returns the total **value** of all coins in this vault,
+// i.e. the realizable total value denominated by GetDenom() if the vault
+// were to liquidate its entire strategies.
+//
+// **Note:** This does not include the tokens held in bank by the module
+// account. If it were to be included, also note that the module account is
+// unblocked and can receive funds from bank sends.
+func (k *Keeper) GetVaultTotalValue(
+	ctx sdk.Context,
+	denom string,
+) (sdk.Coin, error) {
+	enabledVault, found := k.GetAllowedVault(ctx, denom)
+	if !found {
+		return sdk.Coin{}, types.ErrVaultRecordNotFound
+	}
+
+	strategy, err := k.GetStrategy(enabledVault.VaultStrategy)
+	if err != nil {
+		return sdk.Coin{}, types.ErrInvalidVaultStrategy
+	}
+
+	return strategy.GetEstimatedTotalAssets(ctx, enabledVault.Denom)
+}
+
+// GetVaultAccountSupplied returns the supplied amount for a single address
+// within a vault.
+func (k *Keeper) GetVaultAccountSupplied(
+	ctx sdk.Context,
+	denom string,
+	acc sdk.AccAddress,
+) (sdk.Coin, error) {
+	vaultShareRecord, found := k.GetVaultShareRecord(ctx, denom, acc)
+	if !found {
+		return sdk.Coin{}, types.ErrVaultShareRecordNotFound
+	}
+
+	return vaultShareRecord.AmountSupplied, nil
+}
+
+// GetVaultAccountValue returns the value of a single address within a vault
+// if the account were to withdraw their entire balance.
+func (k *Keeper) GetVaultAccountValue(
+	ctx sdk.Context,
+	denom string,
+	acc sdk.AccAddress,
+) (sdk.Coin, error) {
+	totalSupplied, err := k.GetVaultTotalSupplied(ctx, denom)
+	if err != nil {
+		return sdk.Coin{}, err
+	}
+
+	accSupplied, err := k.GetVaultAccountSupplied(ctx, denom, acc)
+	if err != nil {
+		return sdk.Coin{}, err
+	}
+
+	vaultTotalValue, err := k.GetVaultTotalValue(ctx, denom)
+	if err != nil {
+		return sdk.Coin{}, err
+	}
+
+	// Percent of vault account ownership = accountSupply / totalSupply
+	// Value of vault account ownership = percentOwned * totalValue
+	vaultShare := accSupplied.Amount.ToDec().Quo(totalSupplied.Amount.ToDec())
+	shareValueDec := vaultTotalValue.Amount.ToDec().Mul(vaultShare)
```

Comment:
Is it fine to convert to Dec to get a percentage of how much the account owns?

Not entirely sure if this may introduce rounding errors.